### ISSUE TITLE
BUG by reading SpecificCharacterSet

### ISSUE THREAD

**Issue by rtdicomexplorer** [fo-dicom/fo-dicom#1154]
There is a bug when retrieving a dataset, where there are multiple values in = x0008,0x0005 and the first is empty. It tries to recover only the first value without verifying that it is not empty. This leads to selecting a wrong Encoding.
You can try it with the following file: http://www.dclunie.com/images/charset/korean_agfa_infinitt_2008-3.dcm.bz2.
![charSetBugFoDicom](https://user-images.githubusercontent.com/78592813/107019132-9c9f1900-67a1-11eb-98a1-23159b7e59c9.jpg)
  
So i fixed it in DicomDatasetReaderObserver line 145.
![charSetBugFixed](https://user-images.githubusercontent.com/78592813/107019176-a9bc0800-67a1-11eb-909b-583a81812671.jpg)

Best regards.

**Comment by gofal** [comment]
It's not that simple.
It's not a bug, but fo-dicom is not supporting the CodeExtension technique so far.

Standard says that if there is only one value, then this is the encoding to use everywhere. 
If there are more values, then Code Extension Technique is used, that means, that in text there may be Escape Sequences in tags to use the extendet encoding.
If the file has several values and the second one is the ISO 2022 IR 149, then that would mean, that normal strings are encoded in default encoding, but only the strings with the escape sequence should be decoded using ISO 2022 IR 149. So setting the second value as default encoding would be wrong.

Dicom standard further says:

> "Note:
> If the Attribute Specific Character Set (0008,0005) has more than one value and value 1 is empty, it is assumed that value 1 is ISO 2022 IR 6."

So a short fix would rather be, so check if the first value is empty, then to set the encoding to "iso 2022 IR 6".

**Comment by gofal** [comment]
See #43

**Comment by rtdicomexplorer** [comment]
Hi  Reinhard
 thanks for the quick reply.
With this trick all fields are read correctly (maybe it's just a
coincidence).
The problem is that the original information (Patient Name in this case) is
no longer recoverable beause it has been read and saved in the Dataset with
a wrong encoding .. therefore it is useless to use other encodings
afterwards.
I can recovery them just when i read the bytes from the stream...
I tried to send this file via the c-store at
https://www.pixelmed.com/publicdicomserver.html after I retrieved the
information via c-find, and the charset was ISO_IR 192.
So I'm a bit confused as to how to proceed.
Best Regards.

Am Fr., 5. Feb. 2021 um 11:26 Uhr schrieb Reinhard Gruber <
notifications@github.com>:

**Comment by chen-wu** [comment]
If you just want to decode string attributes, then the solution would be simple.

1. Get the buffer of your target attribute.
2. If the 0x00080005 has only one value, just decode the string as it is.
3. If it has multiple values, then get a default encoding by its first value. If it is empty, use IR6.
4. Read byte from the buffer until you get an 0x1B, which means encoding changes from here. Use the default encoding to decode the bytes you have read. 
5. Change the encoding by the escape sequence tokens after 0x1B.
Repeat 4~5 until all bytes are decoded.

Basically you will only need the first value of 0x00080005 to decode string attributes.

**Comment by gofal** [comment]
@chen-wu Thanks for the explanation! It sounds as if you had already implemented it? Would you mind to make a pull request?

**Comment by rtdicomexplorer** [comment]
Hi chen-wu, thank you for your explanation.
I had already adopted a similar solution in my old Dicom-Library (see
attachment).
fo-dicom as it is now does not allow any workaround (unfortunally
..sigh...).
You actually need to change the code. However, I don't think it's that
simple. You need to memorize both encodings and then use them as you said.
Best Regards Michele

Am So., 7. Feb. 2021 um 11:35 Uhr schrieb Reinhard Gruber <
notifications@github.com>:

**Comment by chen-wu** [comment]
@gofal I think it would be difficult. As far as I remembered, the code when fo-dicom try to decode one attribute may not be able to get charset of current dicom object.
I'll try to investigate it again. But if not, I may be able to provide a helper class so someone can perform an additional step to decode string when he want.

@rtdicomexplorer Thanks for your information. I'll see what I can do.

**Comment by gofal** [comment]
@chen-wu it would be simpler to only take a look at version 5 (`FO-DICOM.Full.sln`). I reworked the handling of Encoding there. Instead of storing the Encoding in every single DicomTag now there is one Encoding property per DataSet.
This was necessary because passing the ecoding as parameter in every DicomTag may cause invalid files when you pass different encodings within the same dataset.
So in version 5 it should be more easily to implement encoding extension.

**Comment by chen-wu** [comment]
> @chen-wu it would be simpler to only take a look at version 5 (`FO-DICOM.Full.sln`). I reworked the handling of Encoding there. Instead of storing the Encoding in every single DicomTag now there is one Encoding property per DataSet.
> This was necessary because passing the ecoding as parameter in every DicomTag may cause invalid files when you pass different encodings within the same dataset.
> So in version 5 it should be more easily to implement encoding extension.

OK, I will check it. 
Thank you.

**Comment by kira-96** [comment]
Hello everyone, this is an interesting question. I'm also doing some research recently. I've added support for GBK and GB2312(Fodicom doesn't support these encodings at present), but I don't have any dcm files to test it, although I'm sure it's correct...
see
https://github.com/kira-96/fo-dicom/commit/6a60f7304b09f9f09d1e78c46aabf443a7968292
